commit 941bbfabc3b83bb0b7e1af8e0527e772b58c68d6
Author: mickybart <mickybart@pygoscelis.org>
Date:   Mon Mar 14 17:40:14 2016 +0100

    hw key: camera: support custom settings

diff --git a/services/core/java/com/android/server/policy/PhoneWindowManager.java b/services/core/java/com/android/server/policy/PhoneWindowManager.java
index 1ab44a7..ce88a20 100644
--- a/services/core/java/com/android/server/policy/PhoneWindowManager.java
+++ b/services/core/java/com/android/server/policy/PhoneWindowManager.java
@@ -188,6 +188,11 @@ public class PhoneWindowManager implements WindowManagerPolicy {
     static final int QEMU_HW_MAINKEYS_LAYOUT_DEPRECATED = 0;
     static final int QEMU_HW_MAINKEYS_LAYOUT_50 = 1;
     static final int QEMU_HW_MAINKEYS_MUSIC = 1;
+    
+    //Qemu Hw Camerakey
+    static final int QEMU_HW_CAMERAKEY_DISABLED = 0;
+    static final int QEMU_HW_CAMERAKEY_DEFAULT = 1;
+    static final int QEMU_HW_CAMERAKEY_LPO = 2;
 
     static final int DOUBLE_TAP_HOME_NOTHING = 0;
     static final int DOUBLE_TAP_HOME_RECENT_SYSTEM_UI = 1;
@@ -555,6 +560,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
     //Qemu Hw Mainkeys
     int mQemuHwMainkeysLayout;
     boolean mQemuHwMainkeysMusic;
+    int mQemuHwCamerakey;
     boolean mQemuHwMainkeysIsLongPress;
     boolean mKeysIsLongPress;
 
@@ -1460,6 +1466,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             mQemuHwMainkeysLayout = QEMU_HW_MAINKEYS_LAYOUT_50;
         }
         mQemuHwMainkeysMusic = (SystemProperties.getInt("persist.qemu.hw.mainkeys_music", QEMU_HW_MAINKEYS_MUSIC) == QEMU_HW_MAINKEYS_MUSIC);
+        mQemuHwCamerakey = SystemProperties.getInt("persist.qemu.hw.camerakey", QEMU_HW_CAMERAKEY_DEFAULT);
 
         mAllowTheaterModeWakeFromKey = mContext.getResources().getBoolean(
                 com.android.internal.R.bool.config_allowTheaterModeWakeFromKey);
@@ -5180,18 +5187,22 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             }
 
             case KeyEvent.KEYCODE_CAMERA : {
-                if (!interactive && down) {
-                    isWakeKey = true;
-                    launchCameraApp(keyguardActive);
-                } else if (interactive && down) {
-                    mKeysIsLongPress = false;
-                    scheduleLongPressKeyEvent(event, KeyEvent.KEYCODE_CAMERA);
-                    // Consume key down events of all presses.
-                } else if (interactive) {
-                    mHandler.removeMessages(MSG_CAMERA_LONG_PRESS);
-                    // Consume key up events of long presses only.
-                    if(mKeysIsLongPress) {
+                if (mQemuHwCamerakey != QEMU_HW_CAMERAKEY_DISABLED) {
+                    if (!interactive && down && mQemuHwCamerakey != QEMU_HW_CAMERAKEY_LPO) {
+                        mKeysIsLongPress = false;
+                        isWakeKey = true;
                         launchCameraApp(keyguardActive);
+                    } else if (down) {
+                        mKeysIsLongPress = false;
+                        scheduleLongPressKeyEvent(event, KeyEvent.KEYCODE_CAMERA);
+                        // Consume key down events of all presses.
+                    } else {
+                        mHandler.removeMessages(MSG_CAMERA_LONG_PRESS);
+                        // Consume key up events of long presses only.
+                        if(mKeysIsLongPress) {
+                            isWakeKey = true;
+                            launchCameraApp(keyguardActive);
+                        }
                     }
                 }
                 break;
@@ -5408,6 +5419,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             case KeyEvent.KEYCODE_MEDIA_RECORD:
             case KeyEvent.KEYCODE_MEDIA_FAST_FORWARD:
             case KeyEvent.KEYCODE_MEDIA_AUDIO_TRACK:
+            case KeyEvent.KEYCODE_CAMERA:
                 return false;
         }
         return true;
